File: 0126-word-ladder-ii/0126-word-ladder-ii.py
class Solution(object):
    def findLadders(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: List[List[str]]
        """
        wordSet = set(wordList)
        if endWord not in wordSet:
            return []

        # parent relationships for shortest paths: child -> set of parents
        parents = defaultdict(set)

        # BFS frontier
        level = {beginWord}
        found = False
        L = len(beginWord)

        while level and not found:
            next_level = set()
            # remove all current level words from wordSet to avoid revisiting
            wordSet -= level

            for word in level:
                chars = list(word)
                for i in range(L):
                    orig = chars[i]
                    for c in range(ord('a'), ord('z') + 1):
                        ch = chr(c)
                        if ch == orig:
                            continue
                        chars[i] = ch
                        new_word = ''.join(chars)
                        if new_word in wordSet:
                            # record parent relation
                            parents[new_word].add(word)
                            next_level.add(new_word)
                            if new_word == endWord:
                                found = True
                    chars[i] = orig
            level = next_level

        if not found:
            return []

        # Backtrack all paths from endWord to beginWord using parents
        res = []
        path = [endWord]

        def backtrack(word):
            if word == beginWord:
                res.append(path[::-1])
                return
            for p in parents[word]:
                path.append(p)
                backtrack(p)
                path.pop()

        backtrack(endWord)
        return res

File: 0126-word-ladder-ii/0126-word-ladder-ii.cpp
#include <vector>
#include <string>
#include <unordered_set>
#include <unordered_map>
#include <queue>
using namespace std;

class Solution {
public:
    vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> dict(wordList.begin(), wordList.end());
        vector<vector<string>> result;

        if (!dict.count(endWord)) return result;

        unordered_map<string, vector<string>> parents;
        unordered_set<string> currentLevel{beginWord};
        bool found = false;

        while (!currentLevel.empty() && !found) {
            for (const string &w : currentLevel)
                dict.erase(w);

            unordered_set<string> nextLevel;

            for (const string &word : currentLevel) {
                string temp = word;
                for (int i = 0; i < temp.size(); i++) {
                    char original = temp[i];
                    for (char c = 'a'; c <= 'z'; c++) {
                        temp[i] = c;
                        if (dict.count(temp)) {
                            nextLevel.insert(temp);
                            parents[temp].push_back(word);
                            if (temp == endWord)
                                found = true;
                        }
                    }
                    temp[i] = original;
                }
            }
            currentLevel = nextLevel;
        }

        if (!found) return result;

        vector<string> path{endWord};
        backtrack(endWord, beginWord, parents, path, result);
        return result;
    }

private:
    void backtrack(string word, string beginWord,
                   unordered_map<string, vector<string>>& parents,
                   vector<string>& path,
                   vector<vector<string>>& result) {
        if (word == beginWord) {
            vector<string> temp = path;
            reverse(temp.begin(), temp.end());
            result.push_back(temp);
            return;
        }

        for (const string &p : parents[word]) {
            path.push_back(p);
            backtrack(p, beginWord, parents, path, result);
            path.pop_back();
        }
    }
};


